testlapack/dgeev.go

```diff
@@ -0,0 +1,693 @@
+	// 				tmat := randomSchurCanonical(n, n, rnd)
+	// 				q := randomOrthogonal(n, rnd)
+	// 				tq := zeros(n, n, n)
+	// 				blas64.Gemm(blas.NoTrans, blas.Trans, 1, tmat, q, 0, tq)
+	// 				a := zeros(n, n, n)
+	// 				blas64.Gemm(blas.NoTrans, blas.NoTrans, 1, q, tq, 0, a)
+	// 				ev := make([]complex128, n)
+	// 				for j := 0; j < n; {
+	// 					re := tmat.Data[j*tmat.Stride+j]
+	// 					if j == n-1 || tmat.Data[(j+1)*tmat.Stride+j] == 0 {
+	// 						ev[j] = complex(re, 0)
+	// 						j++
+	// 					} else {
+	// 						im := math.Sqrt(math.Abs(tmat.Data[(j+1)*tmat.Stride+j])) *
+	// 							math.Sqrt(math.Abs(tmat.Data[j*tmat.Stride+j+1]))
+	// 						ev[j] = complex(re, im)
+	// 						ev[j+1] = complex(re, -im)
+	// 						j += 2
+	// 					}
+	// 				}
+	// 				test := dgeevTest{
+	// 					a:      a,
+	// 					evWant: ev,
+	// 					tol:    1e-9,
+	// 				}
+	// 				testDgeev(t, impl, "random", test, jobvl, jobvr, 0, true)
+	// 			}
+	// 		}
+	// 	}
+	// }
+}
+
+func testDgeev(t *testing.T, impl Dgeever, tc string, test dgeevTest, jobvl lapack.JobLeftEV, jobvr lapack.JobRightEV, extra int, optwork bool) {
+	const defaultTol = 1e-13
+	tol := test.tol
+	if tol == 0 {
+		tol = defaultTol
+	}
+
+	a := cloneGeneral(test.a)
+	n := a.Rows
+
+	var vl blas64.General
+	if jobvl == lapack.ComputeLeftEV {
+		vl = nanGeneral(n, n, n)
+	}
+
+	var vr blas64.General
+	if jobvr == lapack.ComputeRightEV {
+		vr = nanGeneral(n, n, n)
+	}
+
+	wr := make([]float64, n)
+	wi := make([]float64, n)
+
+	lwork := max(1, 3*n)
+	if jobvl == lapack.ComputeLeftEV || jobvr == lapack.ComputeRightEV {
+		lwork = max(1, 4*n)
+	}
+	if optwork {
+		work := make([]float64, 1)
+		impl.Dgeev(jobvl, jobvr, n, nil, 1, nil, nil, nil, 1, nil, 1, work, -1)
+		lwork = int(work[0])
+	}
+	work := make([]float64, lwork)
+
+	first := impl.Dgeev(jobvl, jobvr, n, a.Data, a.Stride, wr, wi,
+		vl.Data, vl.Stride, vr.Data, vr.Stride, work, len(work))
+
+	prefix := fmt.Sprintf("Case #%v: n=%v, jobvl=%v, jobvr=%v, extra=%v, optwk=%v",
+		tc, n, jobvl, jobvr, extra, optwork)
+
+	if !generalOutsideAllNaN(vl) {
+		t.Errorf("%v: out-of-range write to VL", prefix)
+	}
+	if !generalOutsideAllNaN(vr) {
+		t.Errorf("%v: out-of-range write to VR", prefix)
+	}
+
+	if first > 0 {
+		t.Log("%v: all eigenvalues haven't been computed, first=%v", prefix, first)
+	}
+
+	// Check that conjugate pair eigevalues are ordered correctly.
+	for i := first; i < n; {
+		if wi[i] == 0 {
+			i++
+			continue
+		}
+		if wr[i] != wr[i+1] {
+			t.Errorf("%v: real parts of %vth conjugate pair not equal", prefix, i)
+		}
+		if wi[i] < 0 || wi[i+1] > 0 {
+			t.Errorf("%v: unexpected ordering of %vth conjugate pair", prefix, i)
+		}
+		i += 2
+	}
+
+	// If the eigenvectors were not computed, check the computed eigenvalues
+	// against provided known eigenvalues.
+	if first > 0 || (jobvl == lapack.None && jobvr == lapack.None) {
+		if test.evWant == nil {
+			return
+		}
+		used := make([]bool, n)
+		for i := first; i < n; i++ {
```

Comment:
The idea is just to take each computed eigenvalue and see if there is an unused known eigenvalue sufficiently close to it. If there is not (i.e., idx == -1), then error. The order of the eigenvalues is unknown and if first > 0, then it is not known which eigenvalues were actually computed, so I can't think of anything better than this. This is also why `i` indexes from first (the index of the first valid eigenvalue), which evWant from 0 (the computed eigenvalue can be any element of `test.evWant`).

Of course, this test assumes that the known eigenvalues are well separated.